File: 169_Majority Element.cpp
class Solution {
public:
    // moore majority vote algorithm
    int majorityElement(vector<int>& nums) {
        int m;
        int count = 0;
        for (int i = 0; i < nums.size(); i++) {
            if (count == 0) {
                m = nums[i];
                count++;
            } else if (nums[i] == m) {
                count++;
            } else
                count--;
        }
        return m;
    }
};

File: go/169.Majority_Element.go
func majorityElement(nums []int) int {
	major := nums[0]
	count := 0
	for _, val := range nums {
		if count == 0 {
			count++
			major = val
			continue
		}

		if val == major {
			count++
		} else {
			count--
		}
	}
	return major
}